File: algorithm-implementation/kosaraju.py
from collections import defaultdict, deque


class Graph:
    def __init__(self, vertices):
        self.V = vertices
        self.graph = defaultdict(list)

    def add_edge(self, u, v):
        self.graph[u].append(v)

    def _dfs(self, v, visited, stack):
        visited[v] = True
        for neighbor in self.graph[v]:
            if not visited[neighbor]:
                self._dfs(neighbor, visited, stack)
        stack.append(v)

    def _transpose(self):
        transposed_graph = Graph(self.V)
        for node in self.graph:
            for neighbor in self.graph[node]:
                transposed_graph.add_edge(neighbor, node)
        return transposed_graph

    def _fill_order(self, visited, stack):
        for i in range(self.V):
            if not visited[i]:
                self._dfs(i, visited, stack)

    def _dfs_util(self, v, visited, component):
        visited[v] = True
        component.append(v)
        for neighbor in self.graph[v]:
            if not visited[neighbor]:
                self._dfs_util(neighbor, visited, component)

    def kosaraju_scc(self):
        stack = deque()
        visited = [False] * self.V

        self._fill_order(visited, stack)

        transposed_graph = self._transpose()

        visited = [False] * self.V
        scc_list = []

        while stack:
            node = stack.pop()
            if not visited[node]:
                component = []
                transposed_graph._dfs_util(node, visited, component)
                scc_list.append(component)

        return scc_list


# Example usage
if __name__ == "__main__":
    g = Graph(5)
    g.add_edge(1, 0)
    g.add_edge(0, 2)
    g.add_edge(2, 1)
    g.add_edge(0, 3)
    g.add_edge(3, 4)

    sccs = g.kosaraju_scc()
    print("Strongly Connected Components:", sccs)

# Strongly Connected Components: [[0, 1, 2], [3], [4]]

File: algorithm/graph/kosaraju.js
class Graph {
  constructor(vertices) {
    this.V = vertices;
    this.graph = Array.from({ length: vertices }, () => []);
  }

  addEdge(u, v) {
    this.graph[u].push(v);
  }

  _dfs(v, visited, stack) {
    visited[v] = true;
    for (const neighbor of this.graph[v]) {
      if (!visited[neighbor]) {
        this._dfs(neighbor, visited, stack);
      }
    }
    stack.push(v);
  }

  _transpose() {
    const transposed = new Graph(this.V);
    for (let u = 0; u < this.V; u++) {
      for (const v of this.graph[u]) {
        transposed.addEdge(v, u);
      }
    }
    return transposed;
  }

  _fillOrder(visited, stack) {
    for (let i = 0; i < this.V; i++) {
      if (!visited[i]) {
        this._dfs(i, visited, stack);
      }
    }
  }

  _dfsUtil(v, visited, component) {
    visited[v] = true;
    component.push(v);
    for (const neighbor of this.graph[v]) {
      if (!visited[neighbor]) {
        this._dfsUtil(neighbor, visited, component);
      }
    }
  }

  kosarajuSCC() {
    const stack = [];
    const visited = Array(this.V).fill(false);

    this._fillOrder(visited, stack);

    const transposed = this._transpose();
    const visitedT = Array(this.V).fill(false);
    const sccList = [];

    while (stack.length > 0) {
      const node = stack.pop();
      if (!visitedT[node]) {
        const component = [];
        transposed._dfsUtil(node, visitedT, component);
        sccList.push(component);
      }
    }

    return sccList;
  }
}

// Example usage
const g = new Graph(5);
g.addEdge(1, 0);
g.addEdge(0, 2);
g.addEdge(2, 1);
g.addEdge(0, 3);
g.addEdge(3, 4);

const sccs = g.kosarajuSCC();
console.log("Strongly Connected Components:", sccs);
